cmd/cluster-image-registry-operator/main.go

```diff
@@ -67,19 +68,14 @@ func runOperator(cmd *cobra.Command, args []string) {
 
 	go metrics.RunServer(metricsPort, stopCh)
 
-	imagePrunerController, err := operator.NewImagePrunerController(cfg)
-	if err != nil {
-		klog.Fatal(err)
-	}
-
-	go imagePrunerController.Run(stopCh)
-
-	controller, err := operator.NewController(cfg)
-	if err != nil {
-		klog.Fatal(err)
-	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		<-stopCh
```

Comment:
Not sure if this applies here but as you are refactoring I have a question about our controllers in general.

What happens when we close this channel? In other words: do controllers stop abruptly or they finish processing whatever they are processing and then return? 

I think that is the first, not the latter. Wouldn't it be better to make them finish whatever they are doing before returning? A WaitGroup of some kind.

Reply:
Today the operator will exit immediately and it shouldn't be a problem (it may be killed or experience power outage at any moment).

https://github.com/kubernetes/kubernetes/blob/eec809aa93863afde30b40f03f8f380f006db794/cmd/kube-controller-manager/app/controllermanager.go#L117
kube-controller-manager doesn't care about it

https://github.com/kubernetes/kubernetes/blob/eec809aa93863afde30b40f03f8f380f006db794/cmd/kube-controller-manager/app/controllermanager.go#L278
when it looses the leadership (afaik SIGTERM also causes that), it also immediately stops.

That said, some controllers in library-go start to use WaitGroups. That might be useful for testing where you don't want to have alive controllers between tests.

Reply:
Interesting. So we could use WaitGroups but we don't want to have them necessarily on this PR?

Reply:
Yes, eventually we may want to use WaitGroups and write unit tests for controllers.